### ISSUE TITLE
Update doc for DesignForExtension

### ISSUE THREAD

**Issue by pbludov** [checkstyle/checkstyle#7608]
https://checkstyle.org/config_design.html#DesignForExtension

Add example for each existing config to show what code is ok and what code cause violation.
Good example is https://checkstyle.org/config_naming.html#AbstractClassName
Look at referenced commits of #5832, to see what and how should be changed in code.

Please play with module/check by means of https://checkstyle.org/cmdline.html to understand how it works.

------------

PLEASE READ https://github.com/checkstyle/checkstyle/wiki/Good-practice-on-how-to-update-examples

**Comment by sulkykookie** [comment]
I'm on it.

**Comment by ImmortalRabbit** [comment]
I am on it.

**Comment by aryaniiit002** [comment]
i want to work on this, can i work on it?

**Comment by mananmehta3** [comment]
If the issue is not assigned I would like to work on it.

**Comment by nick-mancuso** [comment]
@mananmehta3 if “I am on it” comments are more than a month old with no PR, feel free to add “I am on it” comment and proceed.

**Comment by mananmehta3** [comment]
The I am on it message on this issue is around a year old and there is no PR.
So I guess that is okay.

**Comment by Chandan-DK** [comment]
Hi. I just checked https://checkstyle.org/config_design.html#DesignForExtension.
It has examples for all the configs. So, why isn't this issue closed?
Thank you

**Comment by romani** [comment]
https://checkstyle.org/config_design.html#DesignForExtension_Examples

We need example of code for default config.
We need to make all java codes to be same code to show user how only changes in config makes behavior (location of violations) different.

**Comment by Chandan-DK** [comment]
I will work on this issue. 
I'll ask for any clarifications required when I start working on it

**Comment by romani** [comment]
I highly recommend to review other PRs for doc update, example https://github.com/checkstyle/checkstyle/pull/12187 , to see where new contributors have most of misunderstanding and how to make update less time consuming for you.

**Comment by Chandan-DK** [comment]
> We need example of code for default config. We need to make all java codes to be same code to show user how only changes in config makes behavior (location of violations) different.

@romani 
1. You mean that the examples for default config and all these properties (https://checkstyle.org/config_design.html#DesignForExtension_Properties)
should be similar to each other? 
If yes, shall I make changes to the existing examples provided for the properties? Because they are slightly different from each other 

2. Should I include all the default values or just one or two of the default values of ignoredAnnotations property (Before, After, AfterClass, BeforeClass) along with the requiredJavadocPhrase property in the example?

**Comment by romani** [comment]
>shall I make changes to the existing examples provided for the properties?

Yes, we need to change existing java examples.

>Should I include all the default values or just one or two of the default values

We should show in examples all properties that we have in Properties table on doc page

**Comment by Chandan-DK** [comment]
I would like to confirm if these examples are alright.

1. Default Config Example
![image](https://user-images.githubusercontent.com/103727912/198844279-32532a5e-e105-447b-9816-43ad7173a751.png)

2. Example for ignoredAnnotations property with value="Deprecated, Test"
![image](https://user-images.githubusercontent.com/103727912/198844371-6361ed0d-b234-44b3-a377-e96e38d6c0b9.png)

3. Example for requiredJavadocPhrase property with value="This implementation"
![image](https://user-images.githubusercontent.com/103727912/198844446-6a017840-7634-485c-ac8a-be2c53f28287.png)

4. Example for requiredJavadocPhrase property with value="This[\s\S]*implementation"
![image](https://user-images.githubusercontent.com/103727912/198844484-2322fd36-7394-45dd-bc4a-ba82f2dfa888.png)

**Comment by rnveach** [comment]
@Chandan-DK Just pointing out that they are not ok in terms to what @romani specified in https://github.com/checkstyle/checkstyle/issues/7608#issuecomment-1295825128
> We need to make all java codes to be same code to show user how only changes in config makes behavior (location of violations) different.

I would also say that it was mentioned there is existing code. I would take those and try to match them against the properties and wording to get the violations that seem expected to you. If you can't it may be a sign our wording is wrong.

To review actual code you have written, I would recommend put this to a PR and we can actual play with your examples (copy/paste) if we see any issues.

Also, any examples you create, you should display us the CLI output to show they are producing or not producing the violations as you specify. Ex: https://checkstyle.org/report_issue.html#How_to_report_a_bug.3F

**Comment by Chandan-DK** [comment]
@rnveach Ok. I'll make all the code same and submit a PR

**Comment by romani** [comment]
Java code should be identical, only trailing comments are changing to highlight location of violations

**Comment by Chandan-DK** [comment]
After making changes, how do I view .xml files as .html files like this? https://checkstyle.org/config_design.html#DesignForExtension

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations#how-to-generate-website-only

**Comment by Chandan-DK** [comment]
Here's the CLI output of the examples. I hope it's ok?

![image](https://user-images.githubusercontent.com/103727912/198864961-88474f7b-3593-4fcf-b51c-c6124c91c62b.png)

**Comment by nrmancuso** [comment]
@Chandan-DK please share output as code block in markdown, it makes it easy to read and copy paste.

**Comment by Chandan-DK** [comment]
# Default Configuration 

* Code

```Java
public class A {
    
    @BeforeClass
    public int beforeAllTests() { return 1; } // OK, @BeforeClass is ignored by default

    @AfterClass
    public int afterAllTests() { return 1; }  // OK, @AfterClass is ignored by default
    
    @Before
    public int beforeEachTest() { return 1; }  // OK, @Before is ignored by default
    
    @After
    public int afterEachTest() { return 1; }  // OK, @After is ignored by default

    @Test
    public int testSomething() { return 1; }  // OK, @Test is ignored by default 
 
    @Deprecated
    public int someOldMethod() { return 1; }  // Violation, @Deprecated is not ignored by default

    /**
	This method performs some task
    */
    public int someTask() { return 1; } // OK, default requiredJavadocPhrase is in comment

    public int someTask2() { return 1; } // Violation, default requiredJavadocPhrase is not in comment
}
```

* XML Configuration

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
      <module name="DesignForExtension"/>
  </module>
</module>
```



# Ignored Annotations Configuration 

* Code

```Java
public class A {
    
    @BeforeClass
    public int beforeAllTests() { return 1; } // Violation

    @AfterClass
    public int afterAllTests() { return 1; } // Violation
    
    @Before
    public int beforeEachTest() { return 1; } // Violation
    
    @After
    public int afterEachTest() { return 1; } // Violation

    @Test
    public int testSomething() { return 1; } // OK
 
    @Deprecated
    public int someOldMethod() { return 1; } // OK

    /**
	This method performs some task
    */
    public int someTask() { return 1; } // OK, default requiredJavadocPhrase is in comment

    public int someTask2() { return 1; } // Violation, default requiredJavadocPhrase is not in comment
}
```

* XML Configuration

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
      <module name="DesignForExtension">
         <property name="ignoredAnnotations" value="Test, Deprecated"/>
      </module>
  </module>
</module>
```


# Required JavaDoc Phrase Configuration (Example 1) 

* Code

```Java
public class A {
    
    @BeforeClass
    public int beforeAllTests() { return 1; } // OK, @BeforeClass is ignored by default

    @AfterClass
    public int afterAllTests() { return 1; }  // OK, @AfterClass is ignored by default
    
    @Before
    public int beforeEachTest() { return 1; }  // OK, @Before is ignored by default
    
    @After
    public int afterEachTest() { return 1; }  // OK, @After is ignored by default

    @Test
    public int testSomething() { return 1; }  // OK, @Test is ignored by default 
 
    @Deprecated
    public int someOldMethod() { return 1; }  // Violation, @Deprecated is not ignored by default

    /**
	This method performs some task
    */
    public int someTask() { return 1; } // Violation

    /**
	This implementation ....
    */
    public int someTask2() { return 1; } // OK
}
```

* XML Configuration

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
      <module name="DesignForExtension">
         <property name="requiredJavadocPhrase" value="This implementation"/>
      </module>
  </module>
</module>
```



# Required JavaDoc Phrase Configuration (Example 2)

* Code

```Java
public class A {
    
    @BeforeClass
    public int beforeAllTests() { return 1; } // OK, @BeforeClass is ignored by default

    @AfterClass
    public int afterAllTests() { return 1; }  // OK, @AfterClass is ignored by default
    
    @Before
    public int beforeEachTest() { return 1; }  // OK, @Before is ignored by default
    
    @After
    public int afterEachTest() { return 1; }  // OK, @After is ignored by default

    @Test
    public int testSomething() { return 1; }  // OK, @Test is ignored by default 
 
    @Deprecated
    public int someOldMethod() { return 1; }  // Violation, @Deprecated is not ignored by default

    /**
	This method performs some task
    */
    public int someTask() { return 1; } // Violation

    /**
	This cool implementation ....
    */
    public int someTask2() { return 1; } // OK
}
```

* XML Configuration

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
      <module name="DesignForExtension">
          <property name="requiredJavadocPhrase" value="This[\s\S]*implementation"/>
      </module>
  </module>
</module>
```

**Comment by romani** [comment]
Please send Pull Request, it is much easier to discuss items in PR

**Comment by nrmancuso** [comment]
@Chandan-DK The error message is cut off, but it is likely that the xdocs and Javadoc are not in sync. Look at other PRs for documentation updates to see how xml docs and Java docs should match as an example. Also, the command you are running generates the code coverage report, which isn’t needed if we are only updating documentation.

**Comment by romani** [comment]
We are waiting for help to finish code https://github.com/checkstyle/checkstyle/pull/12353

**Comment by MauryanKansara** [comment]
hey @romani I'd like to help but I'm not able to understand what changes do I have to make? Can you please explain what new changes do you want to make?

**Comment by romani** [comment]
Here are examples in code
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java#L224

Content at 
https://github.com/checkstyle/checkstyle/blob/master/src/xdocs/config_design.xml#L174 should be in sync with javadoc.

We need to show users how each property changes results in change is behavior of validation over java code.

Recheck others PR/issues with same name to have more examples of updates

**Comment by bharatsheoran03** [comment]
I want to do work on this issue but I am beginner and unable to understand the code even I had studied on the given links in the comment box. Please help.

**Comment by Rajat2213** [comment]
Hey @romani can you please check this pr

**Comment by HarshSharma0801** [comment]
Can I try this issue please ?

**Comment by romani** [comment]
@HarshSharma0801 , yes, please

**Comment by coderdeadpool** [comment]
@romani I would like to work on this issue.

**Comment by romani** [comment]
@coderdeadpool , we do not assign issues, just make comment "I am on it" and start working on it.

**Comment by HarshSharma0801** [comment]
@coderdeadpool can you share your contact I wanna learn from you as I'am struggling on this one .

**Comment by coderdeadpool** [comment]
@HarshSharma0801  I'm a beginner too, but I'd be happy to share what I know and help you out. You can connect with me on LinkedIn via the link provided in my GitHub profile.

**Comment by HarshSharma0801** [comment]
Thanks a lot sir .

**Comment by HarshSharma0801** [comment]
@coderdeadpool thanks a lot sir .

**Comment by Rishabh7798** [comment]
I want to work on this issue please assign me this task

**Comment by sd9616** [comment]
I am on it

**Comment by romani** [comment]
Fix from @zwwang98 is merged

**Comment by Sloent** [comment]
Hello, I just finished PR on this issue. I am along similar lines to @zwwang98 and could add more examples. Can you turn this question back on?

**Comment by romani** [comment]
Please send PR, it is ok to send PR to closed issue.
